proto/symge/symge-demo.cpp

```diff
@@ -0,0 +1,387 @@
+
+    auto piter = p.begin();
+    auto qiter = q.begin();
+    unsigned pj = piter->first;
+    unsigned qj = qiter->first;
+
+    while (piter!=p.end() || qiter!=q.end()) {
+        unsigned j = std::min(pj, qj);
+        symbol_term t1, t2;
+
+        if (j==pj) {
+            t1 = x*piter->second;
+            ++piter;
+            pj = piter==p.end()? p.npos: piter->first;
+        }
+        if (j==qj) {
+            t2 = y*qiter->second;
+            ++qiter;
+            qj = qiter==q.end()? q.npos: qiter->first;
+        }
+        if (j!=c) {
+            u.push_back({j, define_sym(t1-t2)});
+        }
+    }
+    return u;
+}
+
+// Actual GE reduction.
+// Note: ncol: number of columns before augmentation.
+// Another note: if we're going to preference diagonal pivots
+// (owing to application), then priority queue is overkill.
+template <typename DefineSym>
+void gj_reduce(symmatrix& A, unsigned ncol, DefineSym define_sym) {
+    struct pq_entry {
+        unsigned key; // first non-zero column after pivot column
+        unsigned mincol;
+        unsigned row;
+    };
+
+    struct pq_order_t {
+        bool operator()(const pq_entry& a, const pq_entry& b) const {
+            // The last condition preferences pivots on diagonal elements.
+            return a.key>b.key ||
+                  (a.key==b.key && a.mincol<b.mincol) ||
+                  (a.key==b.key && a.mincol==b.mincol && a.row!=a.key);
+        }
+    };
+
+    std::priority_queue<pq_entry, std::vector<pq_entry>, pq_order_t> pq;
+
+    unsigned n = A.nrow();
+    for (unsigned i = 0; i<n; ++i) {
+        unsigned c = A[i].mincol();
+        if (c<ncol) pq.push({c, c, i});
+    }
+
+    while (!pq.empty()) {
+        pq_entry pick = pq.top();
+        pq.pop();
+
+        unsigned col = pick.key;
+        auto r1 = pick.row;
+
+        while (!pq.empty() && pq.top().key==pick.key) {
+            pq_entry top = pq.top();
+            pq.pop();
+
+            auto r2 = top.row;
+            A[r2] = row_reduce(col, A[r2], A[r1], define_sym);
+
+            unsigned c = A[r2].mincol_after(col);
+            if (c<ncol) pq.push({c, A[r2].mincol(), r2});
+        }
+        unsigned c = A[r1].mincol_after(col);
+        if (c<ncol) pq.push({c, pick.mincol, r1});
+    }
+}
+
+// Validation
+
+template <typename Rng>
+msparse::matrix<double> make_random_matrix(unsigned n, double density, Rng& R) {
+    std::uniform_real_distribution<double> U;
+    msparse::matrix<double> M(n, n);
+
+    for (unsigned i = 0; i<n; ++i) {
+        for (unsigned j = 0; j<n; ++j) {
+            if (i!=j && U(R)>density) continue;
+            double u = U(R);
+            M[i][j] = i==j? n*(1+u): u-0.5;
+        }
+    }
+    return M;
+}
+
+struct symge_stats {
+    unsigned n;
+    unsigned nnz;
+    unsigned nmul;
+    unsigned nsub;
+    unsigned nsym;
+    double relerr;
+};
+
+template <typename Rng>
+symge_stats run_symge_validation(Rng& R, unsigned n, bool debug = false) {
```

Comment:
For the demo, I don't think we need the unit tests, but certainly would for anything going into `modcc`.